terraform/main.tf

```diff
@@ -5,155 +5,227 @@ terraform {
-      disk_type          = "pd-ssd"
-      image_type         = "COS"
-      auto_repair        = true
-      auto_upgrade       = false
-      preemptible        = true
-      initial_node_count = 0
-      # Let's pin this so we don't upgrade each time terraform runs
-      version = "1.19.9-gke.1400"
-    },
-  ]
-
-  node_pools_oauth_scopes = {
-    all = [
-      # FIXME: Is this the minimal?
-      #
-      "https://www.googleapis.com/auth/cloud-platform",
-    ]
+
+  addons_config {
+    http_load_balancing {
+      // FIXME: This used to not work well with websockets, and
+      // cost extra money as well. Let's validate if this is still
+      // true?
+      disabled = true
+    }
+    horizontal_pod_autoscaling {
+      // This isn't used anywhere, so let's turn this off
+      disabled = true
+    }
+    config_connector_config {
+      enabled = var.config_connector_enabled
+    }
+  }
+
+  release_channel {
+    # We upgrade clusters manually so we can manage downtime of
+    # master *and* nodes. When a cluster is in a release channel,
+    # upgrades (including disruptive node upgrades) happen automatically.
+    # So we disable it.
+    channel = "UNSPECIFIED"
+  }
+
+  network_policy {
+    enabled = var.enable_network_policy
   }
 
-  node_pools_labels = {
-    all = {}
+  node_config {
+    # DO NOT TOUCH THIS BLOCK, IT REPLACES ENTIRE CLUSTER LOL
+    service_account = google_service_account.cluster_sa.email
+  }
+}
+
+resource "google_container_node_pool" "core" {
+  name     = "core-pool"
+  cluster  = google_container_cluster.cluster.name
+  project  = google_container_cluster.cluster.project
+  location = google_container_cluster.cluster.location
+
+
+  initial_node_count = 1
+  autoscaling {
+    min_node_count = 1
+    max_node_count = var.core_node_max_count
+  }
+
+  management {
+    auto_repair = true
+    # Auto upgrade will drain and setup nodes without us knowing,
+    # and this can cause outages when it hits the proxy nodes.
+    auto_upgrade = false
+  }
 
-    core-pool = {
-      default-node-pool              = true
-      "hub.jupyter.org/pool-name"    = "core-pool",
+
+  node_config {
+    labels = {
       "hub.jupyter.org/node-purpose" = "core",
       "k8s.dask.org/node-purpose"    = "core"
     }
-    user-pool = {
-      "hub.jupyter.org/pool-name"    = "user-pool"
-      "hub.jupyter.org/node-purpose" = "user",
-      "k8s.dask.org/node-purpose"    = "scheduler"
-    }
-    dask-worker-pool = {
-      "hub.jupyter.org/pool-name" = "dask-worker-pool"
-      "k8s.dask.org/node-purpose" = "worker"
-    }
+    machine_type = var.core_node_machine_type
+    disk_size_gb = 30
+
+    # Our service account gets all OAuth scopes so it can access
+    # all APIs, but only fine grained permissions + roles are
+    # granted via the service account.
+    service_account = google_service_account.cluster_sa.email
+    oauth_scopes = [
+      "https://www.googleapis.com/auth/cloud-platform"
+    ]
```

Comment:
Can you explain more about why you are adding these oauth_scopes in all the nodes?

Reply:
I added a comment!

Reply:
Thanks!